controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorsTest.java

```diff
@@ -90,13 +89,94 @@ public void testEventKey() {
+        }).when(controller).createScope(anyString());
+
+        LinkedBlockingQueue<CompletableFuture<Boolean>> createStreamResponses = new LinkedBlockingQueue<>();
+        LinkedBlockingQueue<CompletableFuture<Void>> createStreamSignals = new LinkedBlockingQueue<>();
+        List<CompletableFuture<Boolean>> createStreamResponsesList = new LinkedList<>();
+        List<CompletableFuture<Void>> createStreamSignalsList = new LinkedList<>();
+        for (int i = 0; i < 4; i++) {
+            CompletableFuture<Boolean> responseFuture = new CompletableFuture<>();
+            CompletableFuture<Void> signalFuture = new CompletableFuture<>();
+            createStreamResponsesList.add(responseFuture);
+            createStreamResponses.add(responseFuture);
+            createStreamSignalsList.add(signalFuture);
+            createStreamSignals.add(signalFuture);
+        }
+
+        // return a future from latches queue
+        doAnswer(x -> {
+            createStreamSignals.take().complete(null);
+            return createStreamResponses.take();
+        }).when(controller).createStream(anyString(), anyString(), any());
+
+        @Cleanup
+        ControllerEventProcessors processors = spy(new ControllerEventProcessors("host1",
+                config, controller, checkpointStore, streamStore, bucketStore,
+                connectionPool, streamMetadataTasks, streamTransactionMetadataTasks,
+                kvtStore, kvtTasks, system, executorService()));
+
+        //Check isReady() method with four possible cases
+        // 1. BootStrap = false and MetadataService = false
+        Assert.assertFalse(processors.getBootstrapCompleted().get());
+        Assert.assertFalse(processors.isMetadataServiceConnected());
+        Assert.assertFalse(processors.isReady());
+        // 2. BootStrap = false and MetadataService = true
+        when(checkpointStore.isHealthy()).thenReturn(true);
+        Assert.assertFalse(processors.isReady());
+
+        // call bootstrap on ControllerEventProcessors
+        processors.bootstrap(streamTransactionMetadataTasks, streamMetadataTasks, kvtTasks);
+        // wait on create scope being called.
+        createScopeSignalFuture.join();
+        createScopeResponseFuture.complete(true);
+        // call create streams now
+        createStreamSignalsList.get(0).join();
+        createStreamSignalsList.get(1).join();
+        createStreamSignalsList.get(2).join();
+        createStreamSignalsList.get(3).join();
+        // complete successfully
+        createStreamResponsesList.get(0).complete(true);
+        createStreamResponsesList.get(1).complete(true);
+        createStreamResponsesList.get(2).complete(true);
+        createStreamResponsesList.get(3).complete(true);
+        AssertExtensions.assertEventuallyEquals(true, () -> processors.getBootstrapCompleted().get(), 10000);
+        // 3. BootStrap = true and MetadataService = false
+        doReturn(false).when(processors).isMetadataServiceConnected();
```

Comment:
Instead of creating a spy from ControllerEventProcessor, could you please mock the CuratorZookeeperClient object and have it return a true or false value and assert isReady based on that?

Reply:
changes made as per discussion